src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -641,15 +609,46 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
 
     val aggregateOption = Aggregators.buildVariantAggregations(this, ec)
 
-    mapAnnotations(finalType, { case (v, va, gs) =>
-      ec.setAll(localGlobalAnnotation, v, va)
+    val localRowType = matrixType.rowType
 
-      aggregateOption.foreach(f => f(v, va, gs))
-      f().zip(inserters)
-        .foldLeft(va) { case (va, (v, inserter)) =>
-          inserter(va, v)
-        }
-    })
+    val newMatrixType = matrixType.copy(metadata = metadata.copy(vaSignature = finalType))
+    val newRowType = newMatrixType.rowType
+
+    copy2(vaSignature = finalType,
+      rdd2 = OrderedRDD2(newMatrixType.orderedRDD2Type,
+        rdd2.orderedPartitioner,
+        rdd2.mapPartitions { it =>
+          val rvb = new RegionValueBuilder()
+
+          it.map { rv =>
+            val ur = new UnsafeRow(localRowType, rv.region, rv.offset)
+
+            val pk = ur.getAs[RPK](0)
+            val v = ur.getAs[RK](1)
+            val va = ur.get(2)
+            val gs = ur.getAs[IndexedSeq[T]](3)
+
+            ec.setAll(localGlobalAnnotation, v, va)
+            aggregateOption.foreach(f => f(v, va, gs))
+            val newVA = f().zip(inserters)
+              .foldLeft(va) { case (va, (v, inserter)) =>
+                inserter(va, v)
+              }
```

Comment:
It's not per genotype, but it seems we're generally always using `while`-loops these days.

Reply:
I believe the snippet from `ec.setAll(...)` to `val newVA = ...` can be the `f` to a `mapAnnotations` call.